packages/events-api/src/http-handler.ts

```diff
@@ -211,12 +214,12 @@ enum ResponseStatus {
   Failure = 500,
 }
 
-type HTTPHandler = (req: IncomingMessage & { body?: string, rawBody?: Buffer }, res: ServerResponse) => void;
+type HTTPHandler = (req: IncomingMessage & { body?: any, rawBody?: Buffer }, res: ServerResponse) => void;
 
 /**
  * A response handler returned by `sendResponse`.
  */
-type ResponseHandler = (err?: (Error & Partial<Omit<CodedError, keyof Error>>) | { status: number }, responseOptions?: {
+export type ResponseHandler = (err?: Error | CodedError | { status: number }, responseOptions?: {
```

Comment:
No, because when `respond` is called with an object that has `status`, that object is *not* a `CodedError`; that is the call signature usually looks like `respond({ status: ResponseStatus.XYZ }, ...);`.

Reply:
I've been able to jog my memory about the design and intention of the API of a response handler function. 🤔💭 

From the [documentation page](https://slack.dev/node-slack-sdk/events-api#custom-responses):

> When the event was handled normally, call respond() with no arguments. If there was an error, call respond(error) with an object that has a status property set to a valid HTTP status code.

For better or worse, `status` is seen as a property of the `error` argument. Yes, we could discuss that in the next major version iteration of this package/API, `status` should be considered one of the possible `responseOptions`, but this is where we are.

In fact, there is some precedent behind this design. In the [`finalhandler` package](https://github.com/pillarjs/finalhandler#finalhandlerreq-res-options), which is used by Express to handle all errors that aren't explicitly handled already, the `error.status` value is used to set the response status code. The main reason I could find for this design choice is so that when a server is running in `NODE_ENV=development`, Express can use the status code the application desired for the error and still display a stack trace in the browser (since the stack trace is provided by `Error` objects). Adhering to this design makes this package more compatible with express response handlers and should make for easier debugging. There are gaps in how this package has implemented this feature (such as only accepting this for 4xx and 5xx status codes), but that's okay.

In conclusion, I think we should define a new `ResponseError` type as an interface that extends `Error`, which has an optional `status?: number;` property. The type of `ResponseHandler` would then change to the following:

```typescript
export type ResponseHandler = (err?: ResponseError, responseOptions?: {
  failWithNoRetry?: boolean;
  redirectLocation?: boolean;
  content?: any;
}) => void;
```

Even though [I said earlier](https://github.com/slackapi/node-slack-sdk/pull/799/files#r305971898) that the first argument should just be an `Error` object, I believe this preserves the same benefit that all `Error`s are a `ResponseError` because `ResponseError` only defines optional properties. In a structural typed language like TypeScript this is a valid design decision, whereas in a nominally typed one we would run into some trouble.

Reply:
@43081j would you like to weigh in on this point?

Reply:
it would be a breaking change (from a types POV) i guess? as you would no longer be able to pass a plain object (`{ status: 400 }`). If we type it as a `ResponseError` (which is `interface ResponseError extends Error { status?: number; }`), you'll have to always pass an error.

That does sound right to me but it seems it'd mean the types no longer permit what the source does.

You could also have `type ResponseError = (Error & {status?: number}) | {status: number}`.

But that is _almost_ like `err?: Error | {status: number}` except the error can have a status.

Reply:
@43081j i think you're right that this would alter the API contract. however, a JavaScript developer using this package would never experience a breakage. We aren't adding any code that uses properties of the argument that are unique to `Error` when a status is provided.

And until this point, every user of this package was using it as a plain JavaScript package. We never offered typings. If third party typings were used and they disagreed with these, we wouldn't know and there's no way for us to align with them.

Therefore I think this small alteration of the contract is safe.